picus/algorithms/lemmas/linear-lemma.rkt

```diff
@@ -3,110 +3,93 @@
-;   #<set: x5 x14 x12> => #<set: x11>
-;   #<set: x5 x14 x11> => #<set: x12>
-;   #<set: x2> => #<set: x6>
-;   #<set: x8> => #<set: x4>
-;   #<set: x1> => #<set: x5>
-;   #<set: x9> => #<set: x3>
-;   #<set: x7> => #<set: x3>
-(define (compute-rcdmap arg-cnsts [arg-indexonly #f] [arg-tag 'default])
-    (cond
-        ; there's a cached version for the given tag, no more computation, just fetch & return
-        [(hash-has-key? :cached-rcdmaps arg-tag) (hash-ref :cached-rcdmaps arg-tag)]
-        ; no cached version, compute, cache and return
-        [else
-            (define res (compute-cdmap arg-cnsts arg-indexonly))
-            (define new-res (make-hash))
-            (for ([key (hash-keys res)])
-                (define vals (hash-ref res key))
-                (for ([val vals])
-                    (when (not (hash-has-key? new-res val)) (hash-set! new-res val (list )))
-                    (hash-set! new-res val (cons key (hash-ref new-res val)))
-                )
-            )
-            ; make immutable
-            (for ([key (hash-keys new-res)])
-                (hash-set! new-res key (list->set (hash-ref new-res key)))
-            )
-            ; store to cache
-            (hash-set! :cached-rcdmaps arg-tag new-res)
-            ; return
-            new-res
-        ]
-    )
-)
+(define (compute-weight-map linear-clauses)
+  (define res (make-hash))
+  (for ([clause (in-list linear-clauses)])
+    (match-define (cons deducible-vars nonlinear-vars) clause)
+    ;; NOTE(sorawee): I think this is not optimal. We should revisit this.
+    (define len-deducible (set-count deducible-vars))
+    (for ([var (in-set deducible-vars)])
+      (hash-update! res var (λ (old) (+ old len-deducible -1)) 0))
+    (for ([var (in-set nonlinear-vars)])
+      (hash-update! res var (λ (old) (+ old len-deducible)) 0)))
+  res)
 
 ; recursively apply linear lemma
-(define (apply-lemma rcdmap ks us)
-    (printf "  # propagation (linear lemma): ")
-    (define new-ks (list->set (set->list ks))) ; (fixme) do this to copy into a mutable set, required by set-* operations
-    (define new-us (list->set (set->list us))) ; (fixme) same as above
-    (define rec? #f) ; whether propagate should be called again
-    (for* ([key (hash-keys rcdmap)])
-        (when (set-empty? (set-subtract key ks))
-            ; all ks are in key, propagate
-            (set! new-ks (set-union new-ks (hash-ref rcdmap key)))
-            (set! new-us (set-subtract new-us (hash-ref rcdmap key)))
-        )
-    )
-    (let ([s0 (set-subtract new-ks ks)])
-        (if (set-empty? s0)
-            (printf "none.\n")
-            (printf "~a added.\n" s0)
-        )
-    )
-    (if (= (set-count ks) (set-count new-ks))
-        ; no updates, return now
-        (values new-ks new-us)
-        ; has updates, call again
-        (apply-lemma rcdmap new-ks new-us)
-    )
-)
\ No newline at end of file
+(define (apply-lemma linear-clauses ks us)
```

Comment:
We should comment what ks and us are supposed to be

Reply:
I think it stands for "known to be unique" signals and unknown signals but maybe we should rename it to be more obvious.